feat(common): add `NgOptimizedImage` directive

AndrewKushnir opened the pull request:
This PR brings the code of the NgOptimizedImage directive into the main branch.

The development happened in a separate `image-directive` branch and all incoming code was reviewed by both Aurora and Angular teams. Before merging the code, we are conducting one more round of code reviews in this PR.

Notes:

- all public APIs are marked as ["Developer Preview"](https://angular.io/guide/releases#developer-preview), we expect to stabilize the API based on the initial feedback and announce the directive as stable later
- the `NgOptimizedImage` directive is marked as standalone, so it can be imported and used directly
- while `NgOptimizedImage` directive is a part of the `@angular/common` NPM package, it would **NOT** be included into the `CommonModule` for some time (we'll consider doing that later once all APIs are stabilized)
- this PR contains documentation for all public API symbols and there will be a followup PR with a guide for angular.io, which would contain more information about the directive (including installation instructions)

// cc @pkozlowski-opensource @kara @khempenius

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
feat(common): add Image directive skeleton (#45627)

This commit adds Image directive skeleton as well as a set of basic tests.

AndrewKushnir committed:
refactor(common): expose `NgImage` directive via a private API

This commit exports experimental `NgImage` directive via a private API of the `@angular/common` package, so that it can be used acorss other Angular packages for testing purposes.

AndrewKushnir committed:
test(common): add a test app for the `NgImage` directive

This commit adds a simple app that uses the `NgImage` directive to simplify further investigation and tests.

kara committed:
refactor(common): change name of directive from NgImage => NgOptimizedImage

We want it to be clear what benefits the image directive
confers over a normal <img> tag, and the `NgImage` name
didn't provide much information. `NgOptimizedImage` makes
it obvious that the new directive is intended to improve
performance.

AndrewKushnir committed:
refactor(common): drop the `loader` input in favor of `IMAGE_LOADER` token

This commit updates the `NgOptimizedImage` directive to drop the `loader` input. Component-specific loaders can still be configured via `IMAGE_LOADER` token and the `loader` input was only useful in case different loaders have to be present in a single template, which doesn't seem to be a common case. We'll be able to re-introduce the input later if needed.

AndrewKushnir committed:
refactor(common): throw an error if `NgOptimizeImage` inputs change

This commit updates the `NgOptimizeImage` directive to add asserts to make sure no inputs are changed after directive initialization.

AndrewKushnir committed:
refactor(common): throw an error if `width` or `height` inputs are missing or invalid

This commit updates the `NgOptimizeImage` directive to add asserts to make sure the `width` and `height` inputs are set and have correct values (numbers only).

AndrewKushnir committed:
feat(common): detect LCP images in `NgOptimizedImage` and assert if `priority` is set

This commit adds extra logic into the `NgOptimizedImage` experimental directive to detect an LCP image and assert whether the `priority` attribute is applied.

AndrewKushnir committed:
refactor(common): mark experimental `NgOptimizedImage` directive as standalone

This commit updates the `NgOptimizedImage` directive as standalone, so it's easier to import it in
 an app (without importing any NgModules). The `NgOptimizedImageModule` is removed as no longer needed.

AndrewKushnir committed:
refactor(common): rename `NgOptimizedImage` directive selector

This commit changes the `NgOptimizedImage` directive selector from `raw-src` to `rawSrc` to better align with the styleguide.

AndrewKushnir committed:
fix(common): detect `data:` and `blob:` inputs in `NgOptimizedImage` directive

This commit updates the `NgOptimizedImage` directive to throw an error when `data:` and `blob:` inputs are used.

khempenius committed:
feat(common): Add image lazy loading and fetchpriority

AndrewKushnir committed:
refactor(common): move `NgOptimizedImage` logic that sets `src` to a different hook

Currently, the logic that sets the `src` on the host `<img>` element is located in the `ngOnChanges` lifecycle hook and guarded by the dev-mode checks that the inputs do not change. However, those checks would be tree-shaken in prod mode and the `src` would be set each time the `ngOnChanges` hook is invoked. This is undesirable and may potentially lead to performance issues.

This commit moves the `src`-related logic to the `ngOnInit` hook instead, which would have the same effect (executed only once, after all inputs are set) and would behave consistently in dev and prod modes.

kara committed:
fix(common): remove default for image width

Previously NgOptimizedImage would default to requesting
an image at the width matching the width attribute in
the image HTML. While this works for width attrs that
match the intrinsic size of the image (e.g. responsive
images or images sized with CSS), this can be a sharp
edge for developers who use the width/height attributes
to set rendered size (i.e. instead of CSS, which one can
do for a fixed size image). In this case, defaulting to
the width attribute for the requested image would mean
requesting always at 1x quality, so screens with a DPR
 of 2+ get an image that is too small. Without a default
request width, the image served by the CDN would likely
be at intrinsic size, so 2x images would look correct.

This PR also updates the NgOptimizedImage sandbox and tests.

kara committed:
fix(common): throw if srcset is used with rawSrc

Currently if you use a static `srcset` with the image directive,
lazy loading no longer works because the image would start to
load before the loading attribute could be set to "lazy". This
commit throws an error if you try to use `srcset` this way.

In a follow-up commit, a new attribute will be added to support
responsive images in a lazy-loading-friendly way.

kara committed:
feat(common): support custom srcset attributes in NgOptimizedImage

This commit adds a `rawSrcset` attribute as a replacement for the
`srcset` attribute. The `srcset` attribute cannot be set statically
on the image directive because it would cause images to start
downloading before the "loading" attribute could be set to "lazy".

Changing the name to `rawSrcset` allows the directive to control
the timing of image loading. It also makes it possible to support
custom loaders for `srcset` file names. Rather than having to repeat
the image origin for each image, the existing `rawSrc` value and
image loader can be composed to generate each full URL string. The
developer must only provide the necessary widths for the `srcset`.

For example, the developer might write:

```markup
<img rawSrc="hermes.jpg" rawSrcset="100w, 200w" ... />
```

with a loader like:

```js
const loader = (config) => `path/${config.src}?w=${config.width}`;
```

and the img tag will ultimately be set up as something like:

```markup
<img src="path/hermes.jpg?w=100" srcset="path/hermes.jpg?w=100 100w, path/hermes.jpg?w=200 200w" .../>
```

AndrewKushnir committed:
test(common): configure e2e testing infrastructure for `NgOptimizedImage` test app

This commit adds the necessary e2e testing infrastructure to the `NgOptimizedImage` test app, so that the test coverage can be extended and extra scenarios can be tested in a browser.

AndrewKushnir committed:
test(common): add e2e tests for LCP check logic of the `NgOptimizedImage` directive

This commit adds e2e tests for the LCP check logic. Those tests are needed to verify the behavior in a real browser (vs relying on a Node environment).

AndrewKushnir committed:
perf(common): monitor LCP only for images without `priority` attribute

This commit optimizes the logic that monitors whether a give image is an LCP element. If an image has the `priority` attribute set, there is no need to include it into monitoring. Also, if we already warned about a particular image (via a `console.warn`) - there is no need to warn again later (to avoid spamming a console).

AndrewKushnir committed:
feat(common): verify that priority images have preconnect links

This commit updates the `NgOptimizedImage` directive to add a logic to detect whether an image, marked with the "priority" attribute has a corresponding `<link rel="preconnect">` tag in the `document.head` for better performance.

kara committed:
fixup! fix(common): remove default for image width

kara committed:
test(common): fix relative URL issue in web tests

The CI failed to run on the last PR, so we didn't catch
that there were web tests failing due to absolute/relative
URL issues. This commit should fix the issue.

AndrewKushnir committed:
test(common): reorg e2e folders for the `NgOptimizedImage` directive tests

This commit updates the e2e folder to separate e2e tests and playground scripts (so it's clear where files are used).

AndrewKushnir committed:
fix(common): sanitize `rawSrc` and `rawSrcset` values in NgOptimizedImage directive

This commit applies a sanitization to values produced by a loader, before they are used for the `src` and `srcset` image element properties.

devversion committed:
ci: temporarily publish snapshots for `image-directive` branch

This commit enables publishing of snapshots for the `image-directive`
feature branch. The artifacts can be accessed with the following steps:

1. Land your change in `image-directive`
2. Go to the corresponding snapshot repo (e.g. `angular/common-builds`)
3. Go to the `image-directive` branch
4. Copy the SHA of the latest commit in that branch
5. Use that SHA to install via NPM. e.g.
   `https://github.com/angular/common-builds.git#SHA`.

AndrewKushnir committed:
refactor(common): create an NgModule for the `NgOptimizedImage` directive

This commit updates the `NgOptimizedImage` directive to drop the `standalone` flag and create a new NgModule which declares and exports it, so that the directive can be used in apps that use pre-v14 version of Angular.

khempenius committed:
feat(common): add loading attr to NgOptimizedImage

Add loading attribute to NgOptimizedImage.

kara committed:
feat(common): add built-in Imgix loader

This commit adds a built-in Imgix loader for the
NgOptimizedImage directive. If you provide the
desired Imgix hostname, an ImageLoader will be
generated with the correct options.

Usage looks like this:

```ts
providers: [
  provideImgixLoader('https://some.imgix.net')
]
```

It sets the "auto=format" flag by default, which
ensures that the smallest image format supported
by the browser is served.

This change also moves the IMAGE_LOADER, ImageLoader,
and ImageLoaderConfig into a new directory that will
be shared by all built-in image loaders.

AndrewKushnir committed:
test: update error codes used in the `NgOptimizedImage` tests

This commit updates the error codes used in the `NgOptimizedImage` tests. The error codes got u[dated recently in a PR that got merged earlier.

AndrewKushnir committed:
feat(common): provide an ability to exclude origins from preconnect checks in NgOptimizedImage

This commit adds an extra logic to add an ability to exclude origins from preconnect checks in NgOptimizedImage by configuring the `PRECONNECT_CHECK_BLOCKLIST` multi-provider.

khempenius committed:
feat(common): add loaders for cloudinary & imagekit

This commit adds loaders for cloudinary and imagekit.

AndrewKushnir committed:
fixup! feat(common): provide an ability to exclude origins from preconnect checks in NgOptimizedImage

AndrewKushnir committed:
test(common): optimize image loader tests of the NgOptimizedImage directive

This commit updates the image loader tests to avoid creating the TestBed environment, since it's not really needed for the tests. Instead, the loader functions are invoked directly and the output is verified.

AndrewKushnir committed:
refactor(common): move LCP image observer to a separate file (#46295)

This is a small refactoring to extract the LCP image observer to a separate file, similar to the preconnect link checker class location.

pkozlowski-opensource committed:
refactor(common): allow loaders to specify the ensurePreconnect option

This commit modifies the provideImgixLoader function to support
an additional ensurePreconnect option. Other loaders could follow
the same pattern to support a similar option as well.

As more loaders get developed we will refactor code to streamline
the loader authoring and enforce consistency.

AndrewKushnir committed:
refactor(common): allow Cloudinary and ImageKit loaders to specify the ensurePreconnect option

This commit modifies the provideCloudinaryLoader and provideImageKitLoader functions to support an additional ensurePreconnect option, similar to the Imgix loader functions.

AndrewKushnir committed:
refactor(common): move loader util functions to a common location

This commit refactors the code to avoid an extra utils file and instead, all extra helpers are moved to a common location.

khempenius committed:
feat(common): add cloudflare loader

Closes #46315

pkozlowski-opensource committed:
refactor(common): remove code duplication in loaders

This change introduces common infrastructure for image
loaders. This removes code duplication and makes it
easier to write loaders - a loader author just needs
to provide a loader config to the URL mapping function.

pkozlowski-opensource committed:
test(common): remove unneeded describe block

This commit removes a nested describe block as it
does't add any value this point. Also moved the
utils tests to the bottom of the file as those are
less important as compared to the other tests.

pkozlowski-opensource committed:
refactor(common): properly configure PRECONNECT_CHECK_BLOCKLIST

Refactoring tests made me realise that we are misinterpreting the
ensurePreconnect option - in fact it should configure the
PRECONNECT_CHECK_BLOCKLIST provider _only_ if the ensurePreconnect
is set to false.

pkozlowski-opensource committed:
refactor(common): remove unnecessary toString conversions

A small refactor that removes explicit toString calls when not needed.

pkozlowski-opensource committed:
refactor(common): unify url error reporting in image loaders

This refactoring pulls url-related error reporting into one place.
It also makes sure that error messages and the related error
reporting logic are tree-shakable.

khempenius committed:
feat(common): add warnings re: image distortion

Checks whether image is visually distorted. Also adds a check to verify that width and height are set to a non-zero number.

AndrewKushnir committed:
refactor(common): add missing format error call in NgOptimizedImage

This commit adds the missing `formatRuntimeError` function to the code that makes sure there is no image distortion.

AndrewKushnir committed:
Revert "fix(common): sanitize `rawSrc` and `rawSrcset` values in NgOptimizedImage directive"

This reverts commit a0fdb8c7e9ec08bd2088614bc354e1155cd2ed85.

The `src` and `srcset` attributes don't pose security threats in modern browser, so sanitization is not really necessary.

AndrewKushnir committed:
refactor(common): throw an error if an absolute URL is passed to Image Loaders

This commit updates the logic of Image Loaders to throw an error in case an absolute URL is provided an an input to the loader function. The loaders can construct URLs based on image paths relative to the bas URL configured for a loader.

pkozlowski-opensource committed:
refactor(common): remove code duplication in image loaders

This commit moves the URL normalization logic in loaders to
the common loader logic.

pkozlowski-opensource committed:
refactor(common): simplify URL construction in image loaders

This commit simplifies the URL construction logic by capturing
the base path and loader config argumnets in one function.

kara committed:
feat(common): add a density cap for image srcsets

With this commit, the NgOptimizedImage directive will throw
a runtime error if it detects that one of the density descriptors
in rawSrcset is higher than 3x. It's generally not recommended
to use density descriptors higher than ~2, as it causes image to
download at very large sizes on mobile screens (thus slowing down
LCP). The density max is set conservatively to 3 in case apps
expect users to zoom in.

In future commits, we may want to throw even at densities >
than 2 and provide a configuration override for the zoom case.

khempenius committed:
feat(common): explain why width/height is required

Update error message text to explain why width & height are required
attributes.

AndrewKushnir committed:
test(common): add parent injector to the `createEnvironmentInjector` calls in tests

This commit updates the `createEnvironmentInjector` calls to pass parent injector, which became required as an argument.

AndrewKushnir committed:
refactor(common): update error messages of NgOptimizedImage directive

This commit updates error messages that the NgOptimizedImage directive throws or logs into a console.

AndrewKushnir committed:
refactor(common): make NgOptimizedImage directive standalone

This commit updates the NgOptimizedImage directive to become standalone and removes no longer needed NgOptimizedImageModule class.

AndrewKushnir committed:
feat(common): define public API surface for NgOptimizedImage directive

This commit annotates a set of symbols exposed by the NgOptimizedImage directive as a public API.

AndrewKushnir committed:
docs: add description and usage notes to the NgOptimizedImage directive

This commit adds the docs for the NgOptimizedImage directive.
As a part of this commit, we also remove an export of directive-related symbols previously exposed as public APIs (i.e. APIs can not be used directly).

josephperrott commented on .circleci/config.yml:
```diff
@@ -71,6 +71,7 @@ var_11: &only_release_branches
     branches:
       only:
         - main
+        - image-directive
```
image-directive is not a releasable branch correct?  Following this PR the branch should no longer be needed

josephperrott commented on packages/common/src/errors.ts:
```diff
@@ -13,8 +13,22 @@
   // NgForOf errors
   NG_FOR_MISSING_DIFFER = -2200,
+
+  // Image directive errors
```
While all of these were defined as part of creating the image directive, many of these are not image specific errors.  Should they be reordered and sectioned appropriately?

AndrewKushnir replied:
The error codes added within this PR are specific to the Image Directive. Some of them may sound too generic though and we can revisit the names as a followup (while keeping error codes as is).

AndrewKushnir commented:
This PR also resolves #42765.

AndrewKushnir committed:
refactor(common): use quotes instead of backticks in missing width/height error message

This commit updates the error message thrown at runtime when an `<img>` doesn't contain width and/or height attributes.

michaelfaith commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
````diff
@@ -0,0 +1,616 @@
+ * ],
+ * ```
+ *
+ * Step 3: update `<img>` tags in templates to use `rawSrc` instead of `rawSrc`.
````
Should this be `rawSrc` instead of `src`?

pkozlowski-opensource replied:
Yep, good catch!

kara commented on packages/common/src/directives/ng_optimized_image/asserts.ts:
```diff
@@ -0,0 +1,25 @@
+import {RuntimeErrorCode} from '../../errors';
+
+/**
+ * Asserts whether an ngDevMode is enabled and throws an error if it's not the case.
```
Nit

```suggestion
 * Asserts whether ngDevMode is enabled and throws an error if it's not the case.
```

kara commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,616 @@
+/***** Assert functions *****/
+
+// Verifies that there is no `src` set on a host element.
+function assertNoConflictingSrc(dir: NgOptimizedImage) {
```
Nit (can be follow-up): seems odd that we have a dedicated asserts.ts file for ng-optimized-image, but none of these asserts are included. Should we move them over?

AndrewKushnir replied:
FYI, capturing this in a planning doc for post-merge cleanup.

kara commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,131 @@
+
+/**
+ * Multi-provider injection token to configure which origins should be excluded
+ * from the preconnect checks. If can either be a single string or an array of strings
```
```suggestion
 * from the preconnect checks. It can either be a single string or an array of strings
```

kara commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,131 @@
+      this.window = win;
+    }
+    if (blocklist) {
+      this.pupulateBlocklist(blocklist);
```
```suggestion
      this.populateBlocklist(blocklist);
```

kara reviewed (approved):
LGTM. Left a few comments for typos / nits that can be addressed now or in follow-ups

kara committed:
fix(common): set bound width and height onto host element

This commit fixes a bug in NgOptimizedImage where
if you bound the width or height attribute (e.g.
`[width]=width`), the attribute would only be
reflected as "ng-reflect-x". The actual "width" or
"height" attribute would not be set on the host
element. This is a problem because the exact named
attribute must be set on the element for the
browser to detect it and use it to reserve space
for the image (and thus prevent CLS).

kara committed:
feat(common): warn if rendered size is much smaller than intrinsic

This commit adds a console warning in development mode
if the ultimate rendered size of the image is much
smaller than the dimensions of the requested image.
In this case, the warning recommends adjusting the
size of the source image or using the `rawSrcset`
attribute to implement responsive sizing.

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+    }
+  }
+
+  private setHostAttribute(name: string, value: string): void {
+    this.renderer.setAttribute(this.imgElement.nativeElement, name, value);
+  }
```
Why use the `renderer` instead of directly calling `setAttribute`?

AndrewKushnir replied:
We've used the `Renderer` abstraction consistently in the code, so this is just another instance of its usage. We can refactor this code when/if we start recommending to bypass the renderer abstraction when interacting with the DOM (as discussed today during the FW sync meeting).

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+      assertValidLoadingInput(this);
+      assertNoImageDistortion(this, this.imgElement, this.renderer);
+      if (this.priority) {
+        const checker = this.injector.get(PreconnectLinkChecker);
```
Why get this from the injector here rather than `inject` it in the constructor?

AndrewKushnir replied:
We do that in dev mode only, so we want to tree-shake-away this code. Injecting it in a constructor and storing it in a class field would leave some extra artifacts (a field itself).

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+      assertNonEmptyWidthAndHeight(this);
+      assertValidLoadingInput(this);
+      assertNoImageDistortion(this, this.imgElement, this.renderer);
+      if (this.priority) {
+        const checker = this.injector.get(PreconnectLinkChecker);
+        checker.check(this.getRewrittenSrc(), this.rawSrc);
```
Refactor this into a function like `assertPriorityPreconnect` ? (passing the checker as a param)

AndrewKushnir replied:
(capturing this in a planning doc for post-merge cleanup)

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+        // Monitor whether an image is an LCP element only in case
+        // the `priority` attribute is missing. Otherwise, an image
+        // has the necessary settings and no extra checks are required.
+        withLCPImageObserver(
+            this.injector,
+            (observer: LCPImageObserver) =>
+                observer.registerImage(this.getRewrittenSrc(), this.rawSrc));
```
I find this coding pattern to be a bit confusing- why not just make it a class method instead of passing an injector/callback?

AndrewKushnir replied:
We want to tree-shake-away this code in prod mode (it's wrapped into `ngDevMode` check). Class properties would remain in the code even if we tree-shake-away function calls.

jelbourn replied:
Ack- I would add a comment that explains the intent, then

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+    }
+    // Must set width/height explicitly in case they are bound (in which case they will
+    // only be reflected and not found by the browser)
+    this.setHostAttribute('width', this.width!.toString());
```
The explicit `toString` wouldn't be necessary if this used directly the native `setAttribute`

AndrewKushnir replied:
Yeah, this is another instance of the Renderer vs direct DOM manipulation topic, which we can refactor once there is a final decision on the Renderer abstraction.

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+                observer.registerImage(this.getRewrittenSrc(), this.rawSrc));
+      }
+    }
+    // Must set width/height explicitly in case they are bound (in which case they will
+    // only be reflected and not found by the browser)
+    this.setHostAttribute('width', this.width!.toString());
+    this.setHostAttribute('height', this.height!.toString());
+
+    this.setHostAttribute('loading', this.getLoadingBehavior());
+    this.setHostAttribute('fetchpriority', this.getFetchPriority());
+    // The `src` and `srcset` attributes should be set last since other attributes
+    // could affect the image's loading behavior.
+    this.setHostAttribute('src', this.getRewrittenSrc());
+    if (this.rawSrcset) {
+      this.setHostAttribute('srcset', this.getRewrittenSrcset());
+    }
```
I would refactor this block into a method like `setImageHostAttributes`

pkozlowski-opensource replied:
Yeh. Extracted to the new `setHostAttributes` function (I'm not sure if including `Image` adds value here.

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+    throw new RuntimeError(
+        RuntimeErrorCode.UNEXPECTED_SRC_ATTR,
+        `${imgDirectiveDetails(dir.rawSrc)} both \`src\` and \`rawSrc\` have been set. ` +
+            `Supplying both of these attributes is not necessary and will break lazy loading. ` +
```
```suggestion
            `Supplying both of these attributes breaks lazy loading. ` +
```
brevity, present tense

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+        RuntimeErrorCode.UNEXPECTED_SRC_ATTR,
+        `${imgDirectiveDetails(dir.rawSrc)} both \`src\` and \`rawSrc\` have been set. ` +
+            `Supplying both of these attributes is not necessary and will break lazy loading. ` +
+            `The NgOptimizedImage directive will set \`src\` itself based on the value of \`rawSrc\`. ` +
```
```suggestion
            `The NgOptimizedImage directive sets \`src\` itself based on the value of \`rawSrc\`. ` +
```
present tense

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+  if (dir.srcset) {
+    throw new RuntimeError(
+        RuntimeErrorCode.UNEXPECTED_SRCSET_ATTR,
+        `${imgDirectiveDetails(dir.rawSrc)} both \`srcset\` and \`rawSrcset\` have been set. ` +
+            `Supplying both of these attributes is not necessary and will break lazy loading. ` +
+            `The NgOptimizedImage directive will set \`srcset\` itself based on the value of ` +
+            `\`rawSrcset\`. To fix this, please remove the \`srcset\` attribute.`);
```
```suggestion
        `${imgDirectiveDetails(dir.rawSrc)} both \`srcset\` and \`rawSrcset\` have been set. ` +
            `Supplying both of these attributes breaks lazy loading. ` +
            `The NgOptimizedImage directive sets \`srcset\` itself based on the value of ` +
            `\`rawSrcset\`. To fix this, please remove the \`srcset\` attribute.`);
```

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+}
+
+// Verifies that the `rawSrc` is not a Blob URL.
+function assertNotBlobURL(dir: NgOptimizedImage) {
```
```suggestion
function assertNotBlobUrl(dir: NgOptimizedImage) {
```
(and any similar functions elsewhere)

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+    throw new RuntimeError(
+        RuntimeErrorCode.INVALID_INPUT,
+        `${imgDirectiveDetails(dir.rawSrc, false)} \`rawSrc\` is a Base64-encoded string ` +
+            `(${rawSrc}). Base64-encoded strings are not supported by the NgOptimizedImage directive. ` +
+            `To fix this, disable the NgOptimizedImage directive for this element ` +
+            `by removing \`rawSrc\` and using a regular \`src\` attribute instead.`);
```
```suggestion
            `(${rawSrc}). NgOptimizedImage does not support Base64-encoded strings. ` +
            `To fix this, disable the NgOptimizedImage directive for this element ` +
            `by removing \`rawSrc\` and using a standard \`src\` attribute instead.`);
```
nit: passive voice, "regular" -> "standard"

(similar for other "...is not supported" text)

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+
+// Verify that none of the listed inputs has changed.
+function assertNoPostInitInputChange(
+    dir: NgOptimizedImage, changes: SimpleChanges, inputs: string[]) {
```
```suggestion
    optimizedImage: NgOptimizedImage, changes: SimpleChanges, inputs: string[]) {
```

pkozlowski-opensource replied:
Actually, I think that this logic is "wrong" and this function shouldn't take the directive instance as an argument. Refactoring in a separate commit.

AndrewKushnir replied:
Also capturing this in a planning doc for post-merge cleanup.

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+}
+
+// Verifies that a specified input is a number greater than 0.
+function assertGreaterThanZeroNumber(
```
```suggestion
function assertGreaterThanZero(
```

jelbourn commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+  if (dir.loading && dir.priority) {
+    throw new RuntimeError(
+        RuntimeErrorCode.INVALID_INPUT,
+        `${imgDirectiveDetails(dir.rawSrc)} the \`loading\` attribute ` +
+            `was used on an image that was marked "priority". ` +
+            `Setting \`loading\` on priority images is not allowed ` +
+            `because these images will always be eagerly loaded. ` +
+            `To fix this, remove the “loading” attribute from the priority image.`);
```
I would probably not care in the directive itself if someone sets `loading="lazy"` on a priority image 

(I guess you could alternatively use a template diagnostic to report that it's unnecessary, but I don't know if it's really worth bothering)

kara replied:
Well, we don't want people to lazy load priority images, right? It's the opposite of the recommended behavior. We could just ignore it, but I think it's worth explaining that it doesn't make sense to do it and we won't do it.

jelbourn replied:
Oh, whoops, I meant to say this the other way around (defaulting to `eager`)

jelbourn commented on packages/core/test/bundling/image-directive/e2e/start-server.js:
```diff
@@ -0,0 +1,17 @@
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
```
Add a file overview comment that explains what this file is for?

jelbourn commented on packages/core/test/bundling/image-directive/e2e/util.ts:
```diff
@@ -0,0 +1,49 @@
+/**
```
I usually take a hard stance against naming files anything like `util`, `helpers`, `common`, etc. Can we set a more specific name here?

jelbourn commented on packages/core/test/bundling/image-directive/e2e/basic/basic.ts:
```diff
@@ -0,0 +1,23 @@
+    useValue: () => 'https://angular.io/assets/images/logos/angular/angular.svg'
+  }],
+})
+export class BasicComponent {
```
Can we name this file and this component more specifically? Something like `SingleOptimizedImage`? 
https://github.com/angular/angular/blob/main/docs/CODING_STANDARDS.md#test-classes-and-examples

AndrewKushnir replied:
(also capturing for post-merge cleanup)

jelbourn reviewed (commented):
I'm most of the way through, but still have a handful of classes to review. I'll get the rest tomorrow

jelbourn commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,131 @@
+  private queryPreconnectLinks(): Set<string> {
+    const preconnectURLs = new Set<string>();
+    const selector = 'link[rel=preconnect]';
+    const links = (this.doc.head?.querySelectorAll(selector) ?? []) as unknown as HTMLLinkElement[];
```
Treating the results of `querySelectorAll` as an array isn't _exactly_ right- you might want to do
```ts
const links = Array.from(document.querySelectorAll('link[preload]')) as HTMLLinkElement[];
```

jelbourn commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -0,0 +1,131 @@
+    const preconnectURLs = new Set<string>();
+    const selector = 'link[rel=preconnect]';
+    const links = (this.doc.head?.querySelectorAll(selector) ?? []) as unknown as HTMLLinkElement[];
+    links.forEach(link => {
```
Prefer `for of`
https://github.com/angular/angular/blob/main/docs/CODING_STANDARDS.md#iteration

jelbourn commented on packages/common/src/directives/ng_optimized_image/util.ts:
```diff
@@ -0,0 +1,64 @@
+/**
```
Same comment as earlier about files call `utils`. It seems like most of this file could be something like `urls.ts` or similar and the directiveDetails function could go into an errors file

jelbourn commented on packages/common/src/directives/ng_optimized_image/util.ts:
```diff
@@ -0,0 +1,64 @@
+export function imgDirectiveDetails(rawSrc: string, includeRawSrc = true) {
+  const rawSrcInfo =
+      includeRawSrc ? `(activated on an <img> element with the \`rawSrc="${rawSrc}"\`) ` : '';
+  return `The NgOptimizedImage directive ${rawSrcInfo}has detected that`;
```
```suggestion
  return `The NgOptimizedImage directive ${rawSrcInfo} has detected that`;
```
missing space

kara replied:
I think the missing space there is deliberate because otherwise if `rawSrcInfo` is '', then there will be two spaces in a row.

jelbourn commented on packages/common/src/directives/ng_optimized_image/util.ts:
```diff
@@ -0,0 +1,64 @@
+// Given a URL, extract the hostname part.
+// If a URL is a relative one - the URL is returned as is.
+export function extractHostname(url: string): string {
+  if (isAbsoluteURL(url)) {
+    const instance = new URL(url);
+    return instance.hostname;
+  }
+  return url;
```
```suggestion
  return isAbsoluteUrl(url) ? (new URL(url)).hostname : url; 
```
optional: this is a matter of personal taste, but I find ternary expressions like this easier to read than the full if-block

jelbourn commented on packages/common/src/directives/ng_optimized_image/util.ts:
```diff
@@ -0,0 +1,64 @@
+    return false;
+  }
+
+  try {
```
Add a comment that explains what error(s) you expect to be caught by this `try`
https://github.com/angular/angular/blob/main/docs/CODING_STANDARDS.md#try-catch

In this case, probably
```ts
// The constructor for `URL` throws an error if passed an invalid URL string.
```

jelbourn reviewed (commented):
Finished first pass through everything, admittedly only really skimmed the tests

kara committed:
refactor(common): clean up util.ts and preconnect_link_checker files

This commit cleans up some stylistic nits and
moves functions out of util.ts into files that
are more specifically named (url.ts and error.ts).

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+ * The `NgOptimizedImage` directive is marked as [standalone](guide/standalone-components) and can
+ * be imported directly.
+ *
+ * Follow the steps below to enable and use the directive:
```
Agreed, we'll refactor the description once the guide is available (will be added in a followup PR).

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+ * 2. Configure a loader that you want to use.
+ * 3. Update the necessary `<img>` tags in templates and replace `src` attributes with `rawSrc`.
+ *
+ * Step 1: import the `NgOptimizedImage` directive.
```
Yeah, the guide is coming in a followup PR, so we'll be able to refactor (and simplify) the content for the `NgOptimizedImage` directive itself.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+  /**
+   * The intrinsic width of the image in px.
+   */
+  @Input()
+  set width(value: string|number|undefined) {
+    ngDevMode && assertGreaterThanZeroNumber(this, value, 'width');
+    this._width = inputToInteger(value);
+  }
```
I don't think there is a particular preference, it was just easier for me to interact with the code where `@Input` is directly above the setter, since this is what gets invoked by the framework when the input is set.

AndrewKushnir committed:
refactor(common): various NgOptimizedImage directive updates

This commit contains a set of small changes based on the feedback received in https://github.com/angular/angular/pull/47082.

pkozlowski-opensource reviewed (approved):
LGTM

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+  private _rewrittenSrc: string|null = null;
+
+  /**
+   * Name of the source image.
```
IMO "name" here works well, as developers are expected to provide only part of the URI (that is - a name, like in a file name). If we would like to be more precise the adequate name would probably be "URI reference", but this is probably too technical / unfamiliar for the users.

gkalpak replied:
FWIW, I also feel `name` is a bit confusing here. I imagine that in many cases this will be a partial path.

pkozlowski-opensource commented on packages/common/src/common.ts:
```diff
@@ -27,3 +27,4 @@ export {PLATFORM_BROWSER_ID as ɵPLATFORM_BROWSER_ID, PLATFORM_SERVER_ID as ɵPL
 export {VERSION} from './version';
 export {ViewportScroller, NullViewportScroller as ɵNullViewportScroller} from './viewport_scroller';
 export {XhrFactory} from './xhr';
+export * from './directives/ng_optimized_image';
```
Agreed, done.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudflare_loader.ts:
```diff
@@ -0,0 +1,34 @@
+ */
+export const provideCloudflareLoader = createImageLoader(
+    createCloudflareURL,
+    ngDevMode ? ['https://<ZONE>/cdn-cgi/image/<OPTIONS>/<SOURCE-IMAGE>'] : undefined);
```
Linked docs for the URL format in the code comment.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudflare_loader.ts:
```diff
@@ -0,0 +1,34 @@
+ */
+
+import {createImageLoader, ImageLoaderConfig} from './image_loader';
+
```
Linked to the product doc: https://developers.cloudflare.com/images/image-resizing/

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudinary_loader.ts:
```diff
@@ -0,0 +1,46 @@
+ *
+ * @param path Base URL of your Cloudinary images
+ * This URL should match one of the following formats:
+ * https://res.cloudinary.com/mysite
+ * https://mysite.cloudinary.com
+ * https://subdomain.mysite.com
```
Linked to the docs in the comment.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+import {isAbsoluteURL, isValidPath, normalizePath, normalizeSrc} from '../util';
+
+/**
+ * Config options recognized by the image loader function.
```
Added references to the `NgOptimizedImage` directive.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+}
+
+/**
+ * Represents an image loader function.
```
Done.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+export type ImageLoader = (config: ImageLoaderConfig) => string;
+
+/**
+ * Noop image loader that does no transformation to the original src and just returns it as is.
+ * This loader is used as a default one if more specific logic is not provided in an app config.
```
Done

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+ */
+export interface ImageLoaderConfig {
+  // Name of the image to be added to the image request URL
+  src: string;
```
Yeh, this would be +1 from me to do `src` -> `name` change. @AndrewKushnir @kara - any objections?

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+  factory: () => noopImageLoader,
+});
+
+export function createImageLoader(
```
Done.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,96 @@
+export function createImageLoader(
+    buildUrlFn: (path: string, config: ImageLoaderConfig) => string, exampleUrls?: string[]) {
+  return function provideImageLoader(
+      path: string, options: {ensurePreconnect?: boolean} = {ensurePreconnect: true}) {
```
The intention here was to have default being `true`, so _not_ providing any option (`undefined`) should result in `true`. The `boolean` type should assure that people don't provide other truthy / falsy values.

pkozlowski-opensource commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -0,0 +1,95 @@
+ * Based on https://web.dev/lcp/#measure-lcp-in-javascript.
+ */
+@Injectable({providedIn: 'root'})
+export class LCPImageObserver implements OnDestroy {
```
No strong opinion here but the `LargestContentfulPaintImageObserver` gets a bit long. @AndrewKushnir @kara any preferences here?

AndrewKushnir replied:
Yeah, I feel the same: the LCP abbreviation becomes quite well established, so we can probably rely on it. Also, since the class is internal (not a part of the public API), we can rename it after we land the initial version (in case there are strong preferences to avoid LCP abbreviation).

pkozlowski-opensource committed:
refactor(common): address review feedback for NgOptimizedImage

Addresses part of the review feedback for the NgOptimizedImage directive.

Akxe commented on packages/common/src/directives/ng_optimized_image/url.ts:
```diff
@@ -0,0 +1,61 @@
+
+// Checks whether a URL is absolute (i.e. starts with `http://` or `https://`).
+export function isAbsoluteUrl(src: string): boolean {
+  return /^https?:\/\//.test(src);
```
`//example.com` is a valid URL too.

Akxe commented on packages/common/src/directives/ng_optimized_image/url.ts:
```diff
@@ -0,0 +1,61 @@
+
+// Checks whether a URL is absolute (i.e. starts with `http://` or `https://`).
+export function isAbsoluteUrl(src: string): boolean {
+  return /^https?:\/\//.test(src);
```
`//example.com/...` is a valid URL too.

neo replied:
so could be like this?

```suggestion
  return /^(?:https?:)?\/\//.test(src);
```

Akxe replied:
The `?:` part is not necessary as the group is not used anyway, but yes. This should do the trick.

AndrewKushnir replied:
Thanks for the feedback, we'll review this and create a followup PR if needed.

pkozlowski-opensource replied:
I'm not sure if we should we encourage protocol-relative URLs given https://www.paulirish.com/2010/the-protocol-relative-url/ 

> Now that SSL is [encouraged for everyone](https://www.eff.org/encrypt-the-web-report) and [doesn’t have performance concerns](https://istlsfastyet.com/), this technique is now an anti-pattern. If the asset you need is available on SSL, then always use the https:// asset.
Allowing the snippet to request over HTTP opens the door for attacks like the [recent Github Man-on-the-side attack](http://www.netresec.com/?page=Blog&month=2015-03&post=China%27s-Man-on-the-Side-Attack-on-GitHub). It’s always safe to request HTTPS assets even if your site is on HTTP, however the reverse [is not true](https://developer.mozilla.org/en-US/docs/Security/MixedContent/How_to_fix_website_with_mixed_content).

Akxe reviewed (changes requested):
This should be fixed. It might cause unexpected results.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+  }
+}
+
+/***** Helpers *****/
```
That makes sense. Kara also mentioned that separately, so I'm capturing this in a planning doc for post-merge cleanup.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+
+// Convert input value to integer.
+function inputToInteger(value: string|number|undefined): number|undefined {
+  return typeof value === 'string' ? parseInt(value, 10) : value;
```
Thanks Jeremy, that makes sense. I was definitely inspired by those functions, but we'd need to align better (and refer to those functions as a source of truth). I've added this item to the post-merge cleanup tracking doc.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/image_loaders/cloudinary_loader.ts:
```diff
@@ -0,0 +1,46 @@
+import {createImageLoader, ImageLoaderConfig} from './image_loader';
+
+/**
+ * Function that generates a built-in ImageLoader for Cloudinary
```
Also capturing this for the followup docs improvement.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,621 @@
+ * throw an error. Such densities are not permitted as they cause image sizes
+ * to be unreasonably large and slow down LCP.
+ */
+export const ABSOLUTE_SRCSET_DENSITY_CAP = 3;
```
cc @kara 

(resolving for now, we can revisit this for all consts post-merge)

AndrewKushnir committed:
refactor(common): minor NgOptimizedImage directive updates

This commit contains a set of small changes based on the feedback received in #47082.

AndrewKushnir commented:
[Presubmit](http://test/OCL:467339956:BASE:467339980:1660356165708:c8911477).

jessicajaniuk reviewed (approved):
reviewed-for: public-api

dylhunn reviewed (approved):
reviewed-for: public-api

Only approving for the public API and developer-facing content. I did *not* review the implementation details.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core 
Reviewed-for: fw-common

alxhub commented on packages/common/src/directives/ng_optimized_image/image_loaders/image_loader.ts:
```diff
@@ -0,0 +1,124 @@
+  const exampleUrlsMsg = exampleUrls.join(' or ');
+  throw new RuntimeError(
+      RuntimeErrorCode.INVALID_LOADER_ARGUMENTS,
+      `Image loader has detected an invalid path (\`${path}\`). ` +
```
I believe the pattern is to use `ngDevMode && 'error message'` to ensure the error messages are tree-shakable in production mode.

pkozlowski-opensource replied:
Actually, functions that throw this runtime error are already guarded with ngDevMode so this is not strictly necessary. But this is a good point - should the error be thrown in the prod mode? I guess so. Changing.

alxhub commented on packages/common/src/directives/ng_optimized_image/url.ts:
```diff
@@ -0,0 +1,61 @@
+
+// Invokes a callback for each element in the array. Also invokes a callback
+// recursively for each nested array.
+export function deepForEach<T>(input: (T|any[])[], fn: (value: T) => void): void {
```
There appear to be multiple impls of this function.

pkozlowski-opensource commented:
Thnx for the review @alxhub - I will merge the "big" PR as I don't want to go through the approvals again. Will send a follow-up cleanup PR so we can review / merge it before the release tomorrow.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 0ca5eb30887b29a455631128f788cb0979115f33.

gkalpak commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,681 @@
+ * RegExpr to determine whether a src in a srcset is using density descriptors.
+ * Should match something like: "1x, 2x".
+ */
+const VALID_DENSITY_DESCRIPTOR_SRCSET = /^((\s*\d(\.\d)?x\s*(,|$)){1,})$/;
```
What about density descriptors with more that one digits (in the integer or fractional part) or omitting the integer part. For example, the current regex will not match the following:
- `1.25x`
- `10x`
- `.9x`

From what I understand by reading [the spec][1], these are valid density descriptors.

[1]: https://html.spec.whatwg.org/multipage/images.html#image-candidate-string

gkalpak commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,681 @@
+ * RegExpr to determine whether a src in a srcset is using density descriptors.
+ * Should match something like: "1x, 2x".
+ */
+const VALID_DENSITY_DESCRIPTOR_SRCSET = /^((\s*\d(\.\d)?x\s*(,|$)){1,})$/;
```
What about density descriptors with more that one digits (in the integer or fractional part) or omitting the integer part. For example, the current regex will not match the following:
- `1.25x`
- `10x`
- `.9x`

From what I understand by reading [the spec][1], these are valid density descriptors.

[1]: https://html.spec.whatwg.org/multipage/images.html#image-candidate-string

kara replied:
> 10x - this one, we intentionally don't support because we have a max density of 3 that we enforce with a warning/error.

> .9x - while technically correct, I don't know of any devices that actually have 0.9x DPR? I don't think we need to support this. Curious what Katie/Alex think on this one though.

> 1.25x - ah, I think we are missing multiple digit support decimals (our existing test just looks for 1.5). Will fix this in a follow-up! Thanks for catching.

khempenius replied:
I don't think we need to support DPRs < 1. Although it's technically possible for devices have that DPR, I think it would imply the existence of a common display format "slightly worse than a CRT display" (which AFAIK is not a thing).

gkalpak replied:
>> 10x
>>
> This one, we intentionally don't support because we have a max density of 3 that we enforce with a warning/error.

That's the point. Without properly detecting `10x` as a valid density descriptor, we throw a confusing error (that the descriptor is not in a valid format) instead of throwing the intended error that the density is too high :smiley:

gkalpak replied:
>> 10x
>>
> This one, we intentionally don't support because we have a max density of 3 that we enforce with a warning/error.

That's the point. Without properly detecting `10x` as a valid density descriptor, we throw a confusing error (that the descriptor is not in a valid format) instead of throwing the intended error that the density is too high :smiley:

Notice in the code below that `isValidDensityDescriptor` will be `false` for `10x`, which will skip the `assertUnderDensityCap()` check and throw the `INVALID_INPUT` error instead:

https://github.com/angular/angular/blob/15552267dd33f203cfaec24892555f7b3f41c33b/packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts#L480-L493

kara replied:
Ohh I see. Yeah, that's definitely not what we want. Will fix that in a follow-up too 👍

gkalpak commented on packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:
```diff
@@ -0,0 +1,681 @@
+ *
+ * To use the **default loader**: no additional code changes are necessary. The URL returned by the
+ * generic loader will always match the value of "src". In other words, this loader applies no
+ * transformations to thr resource URL and the value of the `rawSrc` attribute will be used as is.
```
Typo: `thr resource` --> `the resource`
